src/meta/greedy_load_balancer.cpp

```diff
@@ -1369,8 +1372,74 @@ bool greedy_load_balancer::apply_move(const move_info &move,
                                       /*out*/ migration_list &list,
                                       /*out*/ cluster_migration_info &cluster_info)
 {
-    // TBD(zlw)
-    return false;
+    int32_t app_id = move.pid.get_app_id();
+    rpc_address source = move.source_node, target = move.target_node;
+    if (cluster_info.apps_skew.find(app_id) == cluster_info.apps_skew.end() ||
+        cluster_info.replicas_count.find(source) == cluster_info.replicas_count.end() ||
+        cluster_info.replicas_count.find(target) == cluster_info.replicas_count.end()) {
+        return false;
+    }
+
+    if (cluster_info.apps_info.find(app_id) == cluster_info.apps_info.end()) {
+        return false;
+    }
+    app_migration_info app_info = cluster_info.apps_info[app_id];
+    auto iter1 = app_info.replicas_count.find(source);
+    auto iter2 = app_info.replicas_count.find(target);
+    if (iter1 == app_info.replicas_count.end() || iter2 == app_info.replicas_count.end()) {
+        return false;
+    }
+    app_info.replicas_count[source]--;
+    app_info.replicas_count[target]++;
+    if (app_info.partitions.size() <= move.pid.get_partition_index()) {
+        return false;
+    }
+    auto &pmap = app_info.partitions[move.pid.get_partition_index()];
+    rpc_address primary_addr;
+    for (const auto &kv : pmap) {
+        if (kv.second == partition_status::PS_PRIMARY) {
+            primary_addr = kv.first;
```

Comment:
We can make sure that every partition in app_info must have one primary. Because the app_info in generated in `greedy_load_balancer::get_app_migration_info` and there is a judgement in this function.